lib/recursive-methods.rb

```diff
@@ -1,49 +1,106 @@
 # Authoring recursive algorithms. Add comments including time and space complexity for each method.
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n)
+# Space complexity: O(n)
 def factorial(n)
-    raise NotImplementedError, "Method not implemented"
+    if n < 0
+        raise ArgumentError.new("Negative numbers aren't allowed.")
+    else
+        n == 0 ? 1:n * factorial(n-1)
+    end 
 end
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n)
+# Space complexity: O(n^2)?
 def reverse(s)
-    raise NotImplementedError, "Method not implemented"
+    if s.length <= 1
+        return s
+    else
+        rs = reverse(s[1..-1])
+        rs += s[0]
+        rs
+    end 
 end
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n)
+# Space complexity: O(n^2)?
 def reverse_inplace(s)
-    raise NotImplementedError, "Method not implemented"
+    if s.length <= 1
+        return s
+    else
+        rs = reverse(s[1..-1])
+        rs += s[0]
+        rs
+    end 
 end
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n)
+# Space complexity: O(n)
 def bunny(n)
-    raise NotImplementedError, "Method not implemented"
-end
+    if n == 0
+        return 0
+    else
+        return 2 + bunny(n-1)
+    end 
+end 
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n)
+# Space complexity: O(n)
 def nested(s)
-    raise NotImplementedError, "Method not implemented"
+    b = 0
+    e = s.length - 1
+
+    def valid?(current_beg, current_end, s)
```

Comment:
redefining a method?   It's also a little odd to define a method in a method in Ruby.  That's more of a JavaScript style.